### PR TITLE
[lexical-yjs][lexical-playground] Chore: Add test for collaboration undo bug

### PR DESCRIPTION
**Pull request by mike-atticus**
## Description

Adds a test for the issue https://github.com/facebook/lexical/issues/6614 where using undo in a collaborative session results in the Lexical editor content getting out of sync with the Yjs doc, which then leads to loss of document content. 

### Before

No tests to detect the bugs that lead to unexpected loss of editor content after using undo. 

### After

Playwright test added which uses the side-by-side collaborative playground editors to reproduce the bug after using undo.

Note that this test actually detects two separate bugs:

1. When the left-hand side collaborator clicks undo, the action removes a character of the text typed by the right-hand side collaborator, which shouldn't happen.
2. If you comment out the assertion that checks for bug 1, you'll see that the second paragraph is cleared (including RHS collaborator's text) after LHS collaborator refreshes their page.

### CODE REVIEW

File: packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs [hunk 1]
```diff
@@ -230,4 +230,92 @@ test.describe('Collaboration', () => {
       focusPath: [1, 1, 0],
     });
   });
+
+  test('Undo with two collaborators editing same paragraph', async ({
```
Comment (potatowagon) [inline packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:234]: this is failing on the ci :( 

the failure is reproducible across different envs so i dont think its a flaky failure

Comment (etrepum) [reply packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:234]: That's the point of this PR, to demonstrate that bug

Comment (mike-atticus) [reply packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:234]: Yea I created this test in case it helps others with reproducing the bug and confirming a fix, but it's not something that can be merged on its own. Let me know if I should switch this to a draft PR for now, if that's clearer for PR queue management.

Comment (potatowagon) [reply packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:234]: i see, thanks for filing a detailed issue. if its not meant to be merged, converting to draft or adding a [do not merge] to title would work

Comment (mike-atticus) [reply packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:234]: Title updated 👍

**Comment by james-atticus** [comment]
@etrepum @ivailop7: @mike-atticus and I spent some more time on this and we think we have a fix...

**Our understanding of the issue**

When creating a new paragraph, it starts off with no children.

When a user types a character 'A', we add two things to the paragraph's `CollabElementNode` `_xmlText`: a `YMap` which represents the `TextNode` and the string 'A'. These are added to YJS's undo stack together.

The second user then types character 'B', which adds _only_ the string 'B' to `_xmlText`.

When the first user then clicks Undo, both the `YMap` and 'A' are deleted, however 'B' remains in `_xmlText`. @trueadm added handling for this case in [here](https://github.com/facebook/lexical/pull/1005/files#diff-db90e519357a15ba382479416977bd91e61ae416794f0da2f0c838b5bcdc29cbR188), cleaning up this dangling text when the document is reloaded. However, until that point, Lexical and YJS are out of sync.

**Proposed fix**

The same PR above added [some logic](https://github.com/facebook/lexical/pull/1005/files#diff-db90e519357a15ba382479416977bd91e61ae416794f0da2f0c838b5bcdc29cbR147-R156) for detecting when a text node was deleted leaving dangling text, and tried to merge said text into the previous text node. However, the checks were too strict: the `YMap` can be deleted when `delCount > 1`, as is the case with undo above.

The first change in this PR is fixing the check to properly handle cases when the `YMap` has been deleted from YJS (`if (offset === 0 && length === nodeSize) ...`). The other part is updating the dangling text handling:
* Get the dangling text with `spliceString`
* If there's a preceding text node, move the text to that node
* If there's no preceding text, remove the text from `_xmlText`; i.e.: the same thing as @trueadm's original fix but immediately

Note: this slightly differs from the test case that @mike-atticus originally submitted. At the time, we didn't really understand how Lexical and YJS synced with each other. Once we found @trueadm's PR, we realised the text node is fully deleted from YJS with the undo operation, so it was basically the same issue. Given this [TODO](https://github.com/facebook/lexical/pull/1005/files#diff-db90e519357a15ba382479416977bd91e61ae416794f0da2f0c838b5bcdc29cbR180-R187) about a "proper" fix, we opted to keep it simple and delete the dangling text.

**Next steps**

If Lexical you're are happy with this approach, we'll move the tests over to `Collaboration.test.ts` so that we can add checks on the YJS state and speed up the tests (no need for `sleep`).

**Comment by etrepum** [comment]
The analysis and proposed fix makes sense to me, although I have never done an audit of the collab code so I could be missing something. I think the important things to consider are:

* Is the fix largely backwards compatible (at least such that coherent documents work the same way)? If not, is there another approach that would be more compatible?
* Does it add new edge cases?

The fix sounds simple enough that I think it would be backwards compatible and not cause new trouble, but it's hard to see without at least observing a full e2e collab test suite since there really is a lot going on in here.

Thank you for investigating further!

File: packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs [hunk 2]
```diff
@@ -230,4 +230,163 @@ test.describe('Collaboration', () => {
...
+      page,
+      html`
+        <p
+          class="PlaygroundEditorTheme__paragraph PlaygroundEditorTheme__ltr"
+          dir="ltr">
+          <span data-lexical-text="true">Line 1</span>
+        </p>
+        <p class="PlaygroundEditorTheme__paragraph"><br /></p>
+      `,
+    );
```
Comment (james-atticus) [inline packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:292]: Fails on `main` with:

```
    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 5

      <p
        class="PlaygroundEditorTheme__paragraph PlaygroundEditorTheme__ltr"
        dir="ltr">
        <span data-lexical-text="true">Line 1</span>
      </p>
    - <p class="PlaygroundEditorTheme__paragraph"><br /></p>
    + <p
    +   class="PlaygroundEditorTheme__paragraph PlaygroundEditorTheme__ltr"
    +   dir="ltr">
    +   <span data-lexical-text="true">ord</span>
    + </p>
```

The text is `ord` because Lexical skips a character, as it thinks there's still a `YMap` for the `TextNode`.

File: packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs [hunk 3]
```diff
@@ -230,4 +230,163 @@ test.describe('Collaboration', () => {
...
+    await assertHTML(
+      page,
+      html`
+        <p
+          class="PlaygroundEditorTheme__paragraph PlaygroundEditorTheme__ltr"
+          dir="ltr">
+          <span data-lexical-text="true">normalBOLD</span>
+        </p>
+      `,
+    );
```
Comment (james-atticus) [inline packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:371]: Fails on main with:

```
    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 4

      <p
        class="PlaygroundEditorTheme__paragraph PlaygroundEditorTheme__ltr"
        dir="ltr">
    -   <span data-lexical-text="true">normalBOLD</span>
    +   <span data-lexical-text="true">normal</span>
    +   <strong class="PlaygroundEditorTheme__textBold" data-lexical-text="true">
    +     OLD
    +   </strong>
```

Similar to the test above, a character is dropped because Lexical thinks there's still a `YMap` for the bolded `TextNode`.

File: packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs [hunk 4]
```diff
@@ -230,4 +230,163 @@ test.describe('Collaboration', () => {
...
```
Comment (james-atticus) [inline packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:390]: On main, this `assert` passes for the left editor (because the refresh got it back in sync) but not the right editor (which will only get back in sync when it refreshes).

File: packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs [hunk 5]
```diff
@@ -230,4 +230,163 @@ test.describe('Collaboration', () => {
...
+      page,
+      html`
+        <p
+          class="PlaygroundEditorTheme__paragraph PlaygroundEditorTheme__ltr"
+          dir="ltr">
+          <span data-lexical-text="true">Line 1</span>
+        </p>
+        <p class="PlaygroundEditorTheme__paragraph"><br /></p>
+      `,
+    );
```
Comment (james-atticus) [inline packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:312]: If you comment out the `assert` above, this passes on `main` because of [this handling](https://github.com/facebook/lexical/blob/4fbe7fec1a6f7c44bd828ec11460751f9cbcfe16/packages/lexical-yjs/src/CollabElementNode.ts#L208). The left editor's change is synced over to the right editor.

File: packages/lexical-yjs/src/CollabElementNode.ts [hunk 6]
```diff
@@ -157,21 +157,25 @@ export class CollabElementNode {
...
-              length === nodeSize &&
-              // If the node has no keys, it's been deleted
-              Array.from(node._map.keys()).length === 0
-            ) {
-              // Merge the text node with previous.
-              prevCollabNode._text += node._text;
-              children.splice(nodeIndex, 1);
-            } else if (offset === 0 && delCount === nodeSize) {
-              // The entire thing needs removing
+            if (offset === 0 && length === nodeSize) {
```
Comment (james-atticus) [inline packages/lexical-yjs/src/CollabElementNode.ts:160]: Our observation is that `length < nodeSize` when you're just deleting some text from the node, and `length === nodeSize` if the `YMap` has been deleted. This matches up with the old code's first `if` branch (which also had `Array.from(node._map.keys()).length === 0` but we couldn't see why that was necessary). It also aligns with the second branch `delCount === nodeSize`, where `delCount = min(length, deletionSize)`.

**Review by james-atticus — COMMENTED** [review]
> Is the fix largely backwards compatible?

I believe so, yes. We're only changing logic for processing `delete` deltas from YJS, which shouldn't impact loading existing documents (pretty sure YJS just gives you `insert` events on load). For existing documents that have dangling text at the start of a paragraph, that text will still be deleted from YJS on load by the logic [here](https://github.com/facebook/lexical/blob/4fbe7fec1a6f7c44bd828ec11460751f9cbcfe16/packages/lexical-yjs/src/CollabElementNode.ts#L208).

> Does it add new edge cases?

Not that we could think of. The change simplifies the overall deletion-handling logic; previously there were two branches that called `children.splice(nodeIndex, 1)`, each in very specific circumstances (either `delCount === 1` or `delCount === nodeSize)`. Now it's "if YMap is deleted, _always_ `splice` the CollabTextNode out", regardless of `delCount`, with some logic to either move dangling text to another TextNode or clean it up from YJS (`delCount` is used to determine what part of the text is left over).

We've pushed the code for the proposed fix as well as the two E2E tests that were failing before the change, covering the two paths for handling of dangling text. Are these what you meant by "full e2e collab test suite"? In case it's helpful for understanding how we got to these tests, I've added some comments inline highlighting where they're failing on `main`.

File: packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs [hunk 7]
```diff
@@ -230,4 +230,163 @@ test.describe('Collaboration', () => {
...
   });
+
+  test('Remove dangling text from YJS when there is no preceding text node', async ({
+    isRichText,
+    page,
+    isCollab,
+    browserName,
+  }) => {
+    // test.skip(!isCollab || IS_MAC);
+    test.skip(!isCollab);
```
Comment (james-atticus) [inline packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:241]: Ditto below

```suggestion
    test.skip(!isCollab || IS_MAC);
```

**Comment by ivailop7** [comment]
Pretty confident this is backwards-compatible too. It doesn't modify the stored Ydoc and how to interpret it. I'm happy with the change and tests.

File: packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs (hunk 7 above)
Comment (etrepum) [reply packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:241]: Is there a reason this should be skipped on Mac? I am not seeing any platform specific code here, I think what this will do is make it harder for contributors to test locally on a Mac because some of the tests just won't run.

Comment (james-atticus) [reply packages/lexical-playground/__tests__/e2e/Collaboration.spec.mjs:241]: No particular reason, just pattern matching existing tests. Works fine on Mac so happy to leave as is.

**Review by etrepum — APPROVED** [review]
Given that all the test suites seem to be passing now I don't see any reason not to approve and merge this fix with additional tests. Thank you very much for doing all the hard work on this one!
